### PR TITLE
Overhaul of CI infrastructure

### PR DESCRIPTION
**Pull request by soenkeliebau**
This is a overhaul of our overall CI structure to build, test and deploy artifacts. It moves all build jobs to a single workflow so we can properly declare dependencies between them.

Also removes the Generate Manifests action which adds a commit to all PRs. This is replaced by a job which regenerates charts and manifests and then fails the workflow if `git diff` shows any changes. This is based on the assumption that in that case obsolete crds, charts and/or manifests were committed.

fixes https://github.com/stackabletech/issues/issues/156

Also adds various fixes:

- update helm & cargo deny actions
- remove dependabot and add renovate
- change bors to wait for an approving review
- remove (but not delete) helm roles.yaml so those can be managed in the operators
- fix shebang in generate manifests
- fix weird cp behavior for product config in makefile (subdirectory 'config-spec' should be removed now)
- add reviewdog workflow to run various linters on pull requests
- add udeps job to find unused dependencies

I have created an example PR that should match what will be generated by this once it is merged here: https://github.com/stackabletech/zookeeper-operator/pull/388

### CODE REVIEW

File: repositories.yaml [hunk 1]
```diff
@@ -83,3 +83,10 @@ repositories:
 
 retired_files:
   - .github/workflows/helm_tests.yml
+  - .github/workflows/pr_generate_manifests.yml
+  - .github/workflows/publish_main_artifacts.yml
+  - .github/workflows/publish_pr_artifacts.yml
+  - .github/workflows/publish_release_artifacts.yml
```
Comment (lfrancke) [inline repositories.yaml:89]: Could you add the extra cargo version py again here?

File: template/.github/workflows/build.yml.j2 [hunk 2]
```diff
@@ -0,0 +1,303 @@
...
+  RUSTDOCFLAGS: "-D warnings"
+  RUST_LOG: "info"
+  PRODUCT_NAME: {[ operator.product_string }]
+  DEV_REPO_HELM_URL: https://repo.stackable.tech/repository/helm-dev
+  TEST_REPO_HELM_URL: https://repo.stackable.tech/repository/helm-test
+  STABLE_REPO_HELM_URL: https://repo.stackable.tech/repository/helm-stable
+
+jobs:
+  # Identify unused dependencies
+  udeps:
```
Comment (lfrancke) [inline template/.github/workflows/build.yml.j2:29]: Now that it works and is fast enough I believe we can make it mandatory as well.
And just for consistency: You call all the other ones run_foo so this should be run_udeps

File: template/.github/workflows/build.yml.j2 [hunk 3]
```diff
@@ -0,0 +1,303 @@
...
+      - uses: actions-rs/toolchain@v1.0.7
+        with:
+          profile: minimal
+          toolchain: stable
+          components: rustfmt
+          override: true
+      - name: Install requirements for version tool
+        if: ${{ github.event_name == 'pull_request' }}
+        run: pip install -r python/requirements.txt
+        # This step checks if the current run was triggered by a push to a pr (or a pr being created).
```
Comment (lfrancke) [inline template/.github/workflows/build.yml.j2:280]: Newlines between the steps would make it easier to see here especially with the comments but that's a nit

File: template/.github/workflows/reviewdog.yaml [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+
+  detect-secrets:
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v2
+      - uses: reviewdog/action-detect-secrets@master
+        with:
+          github_token: ${{ secrets.github_token }}
+
+  flake8:
```
Comment (lfrancke) [inline template/.github/workflows/reviewdog.yaml:29]: Could you add another file to the template to add to each repos root dir?

`.flake8`
```
[flake8]
ignore = E111,E501,E114
```

**Review by lfrancke — CHANGES_REQUESTED** [review]
minor stuff, mostly my fault :)

File: repositories.yaml (hunk 1 above)
Comment (soenkeliebau) [reply repositories.yaml:89]: done

File: template/.github/workflows/build.yml.j2 (hunk 2 above)
Comment (soenkeliebau) [reply template/.github/workflows/build.yml.j2:29]: done

File: template/.github/workflows/build.yml.j2 (hunk 3 above)
Comment (soenkeliebau) [reply template/.github/workflows/build.yml.j2:280]: done

File: template/.github/workflows/reviewdog.yaml (hunk 4 above)
Comment (soenkeliebau) [reply template/.github/workflows/reviewdog.yaml:29]: done

File: template/.github/workflows/build.yml.j2 [hunk 5]
```diff
@@ -0,0 +1,308 @@
...
+    steps:
+      - uses: actions/checkout@v2.4.0
+      - uses: actions-rs/toolchain@v1.0.7
+        with:
+          profile: minimal
+          toolchain: stable
+          override: true
+      - uses: Swatinem/rust-cache@v1.3.0
+        with:
+          key: udeps
```
Comment (soenkeliebau) [inline template/.github/workflows/build.yml.j2:43]: Do these keys need to be specific to the repos? If so we should include the name of the operator maybe?

**Review by lfrancke — DISMISSED** [review]
I'm not sure why you reindented the python file but I'm fine. I look forward to have these changes merged!

File: template/.github/workflows/build.yml.j2 (hunk 5 above)
Comment (lfrancke) [reply template/.github/workflows/build.yml.j2:43]: no, they don't

You can remove the cache for now. It won't be used: https://github.com/Swatinem/rust-cache/issues/9#issuecomment-1030048800

**Comment by soenkeliebau** [comment]
> I'm not sure why you reindented the python file but I'm fine. I look forward to have these changes merged!

I ran pylint offline on the file and it complained about every single line. Brief googling seemed to confirm 4 space-indentation as the norm for python ..

**Comment by lfrancke** [comment]
Ah! One more thing. Could you add .flake8 and .pylintrc (https://github.com/stackabletech/operator-templating/blob/main/template/.pylintrc) also to the main directory of the operator-templating repo?

I think that explains your last commit

**Comment by lfrancke** [comment]
Yeah 4 is a default but someone started with 2 here and I liked it better as well (consistent with our other stuff) so I decided to adapt pylint but I don't really have a super strong opinion.

With an indentation of 4 it'll conflict with the pylintrc file in the repos though.
So we should either change the pylintrc file _or_ revert your formatting commit. I don't care which option you choose :)

**Comment by soenkeliebau** [comment]
reverted it to 2 space indentation

**Comment by soenkeliebau** [comment]
also added the linter config to templating

**Review by lfrancke — APPROVED** [review]

File: template/.github/workflows/build.yml.j2 (hunk 5 above)
Comment (lfrancke) [reply template/.github/workflows/build.yml.j2:43]: Ignore me, sorry. This was my mistake. Cache will be used for udeps, it won't be used for cargo-deny.

Comment (soenkeliebau) [reply template/.github/workflows/build.yml.j2:43]: discussed directly with Lars:
Cache is not enabled for cargo-deny, which this comment refers to. For udeps it is used and enabled.

**Pull request merged**